admin/cleanup.py

```diff
@@ -0,0 +1,379 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+"""
+Utilities for cloud resource cleanup.
+"""
+from datetime import datetime, timedelta
+import json
+import sys
+from uuid import UUID
+import yaml
+from yaml.parser import ParserError
+
+from characteristic import attributes
+
+from dateutil.parser import parse as parse_date
+from dateutil.tz import tzutc
+
+from libcloud.compute.providers import get_driver, Provider
+
+from twisted.internet.defer import succeed
+from twisted.python.filepath import FilePath
+from twisted.python.log import err
+from twisted.python.usage import Options, UsageError
+
+# Marker value defined by ``flocker.testtools.cluster_utils.MARKER``.  This
+# should never change and should always identify test-created clusters.
+from flocker.testtools.cluster_utils import MARKER
+
+
+def get_creation_time(node):
+    """
+    Get the creation time of a libcloud node.
+
+    Rackspace and EC2 store the information in different metadeta.
+
+    :return: The creation time, if available.
+    :rtype: datetime or None
+    """
+    date_string = node.extra.get("created", node.extra.get("launch_time"))
+    if date_string is None:
+        return None
+    else:
+        return parse_date(date_string)
+
+
+def get_rackspace_driver(rackspace):
+    """
+    Get a libcloud Rackspace driver given some credentials and other
+    configuration.
+    """
+    rackspace = get_driver(Provider.RACKSPACE)(
```

Comment:
:question: One question on the approach -- I think it's probably more important for us to have something in place rather than do everything precisely correct so more just take this into consideration rather than go off and implement it.

So we already call `get_driver(Provider.RACKSPACE)` in `flocker.provision._rackspace.rackspace_provisioner`. This suggested to me that we might actually want to extend the provisioner to have whatever functionality we need to clean up volumes rather than do a separate libcloud interaction here.

Pros:

- Makes it obvious that you have to add this functionality when you add a new cloud (For instance, I know for GCE I had to implement ``IProvisioner``, I didn't know that for GCE I had to find this script and add GCE support).

- We might need to ensure that the way we query for volumes is consistent with how we are storing the volumes. (This might even suggest that we need a new endpoint in the IBlockDeviceAPI). For instance, for GCE to get the cluster_id you have to query the description, parse it as JSON, and extract a specific key. We might want to change that API to add a specific tag for the cluster. That requires logic in the cleanup code that must be in-sync with the implementation of the IBlockDeviceAPI. Seems like it be nice if that was behind the same interface.

Cons:

- This is implemented.

- `libcloud` already gives us a common interface to multiple clouds, why are we just adding provisioner layers plumbing through to libcloud?

- Really we to be able to query for all ways that we have ever stored volume data, that sort of code might be better served in isolated cleanup code rather than confusing our `pristine` driver implementations.

---------------------

Something else occurred to me, what if we just extend the provisioner interface to have a method that gets a libcloud driver?  You could create the drivers by creating all provisioners in https://github.com/ClusterHQ/flocker/blob/master/flocker/provision/__init__.py#L16, and then we wouldn't even have to change this file to get support for other clouds.